File: 0045-jump-game-ii/0045-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        end = len(nums) - 1
        num_jumps = 0
        curr_max_range = 0
        next_endpoint = 0

        for i in range(end):
            curr_max_range = max(curr_max_range, i+nums[i])

            if curr_max_range >= end:
                num_jumps+=1
                break
        
            if i == next_endpoint:
                num_jumps+=1
                


                next_endpoint = curr_max_range

        return num_jumps

File: 0045-jump-game-ii/0045-jump-game-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var jump = function(nums) {
    if (nums.length <= 1) return 0;
    let numJumps = 0;
    let currReach = 0;
    let maxReach = 0;

    for (let i = 0; i < nums.length; i++) {
        currReach = Math.max(currReach, nums[i] + i);

        if (currReach >= nums.length - 1) {
            numJumps += 1;
            break;
        };

        if (maxReach === i) {
            maxReach = currReach;
            numJumps += 1;
        };
    };

    return numJumps;
};